hazelcast/src/test/java/com/hazelcast/serialization/SerializationBenchmark.java

```diff
@@ -0,0 +1,166 @@
+    Serializer identifiedDataSerializableSerializer;
+
+    Portable portable;
+    Serializer portableSerializer;
+
+    Object object;
+    Serializer streamSerializer;
+
+    @Setup
+    public void setup() {
+        serializable = SerializableObject.object();
+        serializableSerializer = new Serializer(SerializableObject.serializationService());
+
+        externalizable = ExternalizableObject.object();
+        externalizableSerializer = new Serializer(ExternalizableObject.serializationService());
+
+        dataSerializable = DataSerializableObject.object();
+        dataSerializableSerializer = new Serializer(DataSerializableObject.serializationService());
+
+        identifiedDataSerializable = IdentifiedDataSerializableObject.object();
+        identifiedDataSerializableSerializer = new Serializer(IdentifiedDataSerializableObject.serializationService());
+
+        portable = PortableObject.object();
+        portableSerializer = new Serializer(PortableObject.serializationService());
+
+        object = StreamObject.object();
+        streamSerializer = new Serializer(StreamObject.serializationService());
+    }
+
+    @Benchmark
+    public Object[] serializable() throws Exception {
+        byte[] serialized = serializableSerializer.serialize(serializable);
+        return serializableSerializer.deserialize(serialized);
+    }
+
+    @Benchmark
+    public Object[] externalizable() throws Exception {
+        byte[] serialized = externalizableSerializer.serialize(externalizable);
+        return externalizableSerializer.deserialize(serialized);
+    }
+
+    @Benchmark
+    public Object[] dataSerializable() throws Exception {
+        byte[] serialized = dataSerializableSerializer.serialize(dataSerializable);
+        return dataSerializableSerializer.deserialize(serialized);
+    }
+
+    @Benchmark
+    public Object[] identifiedDataSerializable() throws Exception {
+        byte[] serialized = identifiedDataSerializableSerializer.serialize(identifiedDataSerializable);
+        return identifiedDataSerializableSerializer.deserialize(serialized);
+    }
+
+    @Benchmark
+    public Object[] portable() throws Exception {
+        byte[] serialized = portableSerializer.serialize(portable);
+        return portableSerializer.deserialize(serialized);
+    }
+
+    @Benchmark
+    public Object[] stream() throws Exception {
+        byte[] serialized = streamSerializer.serialize(object);
+        return streamSerializer.deserialize(serialized);
+    }
+
+    private static class Serializer {
+
+        private final InternalSerializationService serializationService;
+
+        Serializer(InternalSerializationService serializationService) {
+            this.serializationService = serializationService;
+        }
+
+        byte[] serialize(Object object) throws IOException {
+            try (BufferObjectDataOutput output = serializationService.createObjectDataOutput()) {
+                for (int i = 0; i < BATCH_SIZE; i++) {
```

Comment:
With writing to the `BufferObjectDataOutput` from a loop, we incorporate the cost of the used `BufferObjectDataOutput` implementation into the serialization benchmark result. I guess the benchmark uses `ByteArrayObjectDataOutput` behind the scenes, which resizes itself internally on demand, making the results biased in favor of the strategies allocating less. 

I wonder if we should use `@Measurement(..., batch-size=BATCH_SIZE)` instead of the for loop, with maybe reusing `BufferObjectDataOutput` between the JMH operations, since - AFAIK - we don't create new instances for every serialization call. 

On the other hand, the allocation rate is an important dimension of the serializers. We can use `GCProfiler`, which includes the normalized allocation rate in the results, thus we can include the throughput and the allocation rate of the serializers in the result separated:
```
    public static void main(String[] args) throws Exception {
        Options options = new OptionsBuilder()
                .include(SerializationBenchmark.class.getSimpleName())
                .addProfiler(GCProfiler.class)
                .resultFormat(JSON)
                .build();

        new Runner(options).run();
    }
```
The JSON format can be handy for visualizing the results with https://jmh.morethan.io/.


Maybe @pveentjer can add on this.

Reply:
@gierlachg 
I would use serializatioService.toData/toObject methods directly. 
toData creates `ObjectDataOutput`  but pools them. So the cost of it will be gone after the warmup. 
And when it is done there will be no need to do batching (BATCH_SIZE) here.

Reply:
@blazember I jut tried https://jmh.morethan.io/ it seems very useful especially to compare multiple results.

Reply:
@gierlachg Did you have a chance to take a look at the cost of the serializer creation/allocation? Sancar's suggestion can also be a way to deal with it.